In [1]:
import numpy as np
import pandas as pd
import warnings
with warnings.catch_warnings():
    warnings.filterwarnings("ignore", category=DeprecationWarning)
    import statsmodels.api as sm
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('income.csv')
df.head()

Education     Income
0  10.000000  26.658839
1  10.401338  27.306435
2  10.842809  22.132410
3  11.244147  21.169840
4  11.645485  15.192634

1. The mean, variance, and standard deviation of the education column

The mean, variance, and standard deviation of the income column

The covariance of the education and the income

The correlation of the education and the income
You can use the pandas mean, var, std, cov, and corr methods to do this.

In [3]:
print('mean of education: ' + str(df['Education'].mean()) + '\n' +
     'variance of education: ' + str(df['Education'].var()) + '\n' +
     'std of education: '+ str(df['Education'].std()))

mean of education: 16.0
variance of education: 13.27119615477568
std of education: 3.642965296949132


In [4]:
print('mean of income: ' + str(df['Income'].mean()) + '\n' +
     'variance of income: ' + str(df['Income'].var()) + '\n' +
     'std of income: '+ str(df['Income'].std()))

mean of income: 50.145469306666676
variance of income: 446.9652465894925
std of income: 21.14155260593442


In [5]:
print('covariance:')
df.cov()

covariance:


Education      Income
Education  13.271196   74.311836
Income     74.311836  446.965247

In [6]:
print('correlation:')
df.corr()

correlation:


Education    Income
Education   1.000000  0.964864
Income      0.964864  1.000000

2. Using statsmodels (code similar to that given in lecture), run a regression that uses the education of a person to predict their income. What is your beta (slope of the regression line)? Your r squared? The 95% confidence interval of the beta? Are you confident that beta is greater than zero?

In [7]:
x1 = sm.add_constant(df['Education'], prepend=False)
mod = sm.OLS(df['Income'], x1)
res = mod.fit()
print(res.summary())

                            OLS Regression Results                            
Dep. Variable:                 Income   R-squared:                       0.931
Model:                            OLS   Adj. R-squared:                  0.928
Method:                 Least Squares   F-statistic:                     377.6
Date:                Fri, 05 May 2023   Prob (F-statistic):           8.63e-18
Time:                        23:28:05   Log-Likelihood:                -93.500
No. Observations:                  30   AIC:                             191.0
Df Residuals:                      28   BIC:                             193.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Education      5.5995      0.288     19.431      0.0

### Beta is 5.5995, and r-squared is 0.931, 95% conidence interval is between 5.009 and 6.190. I am confident that beta is greater than zero, because P>|t| is almost 0, which means that we can reject the null hypothesis

3. Using the regression model you created for the last question, calculate the predicted income and add it as a column to your data frame. Then add the residual, which is the difference between the actual and predicted values, to your dataframe as well. 
Compute and print the mean of your predicted income. It will be the same as another mean that you computed before. What is it the same as?
What is the mean of your residuals? Calculate and print it.

In [8]:
#5.5995*df['Education']-39.4463
df['predicted income'] = res.predict(x1)
df.head()

Education     Income  predicted income
0  10.000000  26.658839         16.548572
1  10.401338  27.306435         18.795856
2  10.842809  22.132410         21.267869
3  11.244147  21.169840         23.515153
4  11.645485  15.192634         25.762437

In [9]:
df['residual'] = res.resid
df.head()

Education     Income  predicted income   residual
0  10.000000  26.658839         16.548572  10.110267
1  10.401338  27.306435         18.795856   8.510579
2  10.842809  22.132410         21.267869   0.864542
3  11.244147  21.169840         23.515153  -2.345312
4  11.645485  15.192634         25.762437 -10.569803

In [10]:
df['predicted income'].mean()

50.145469306666655

In [11]:
df['Income'].mean()

50.145469306666676

### We can see that the mean of the predicted income is the same as the mean of the original income, 50.145

In [12]:
print('mean of residual:' + str(df['residual'].mean()))

mean of residual:2.8184861851817308e-14


4. Using your new column of residuals, calculate and print out the RSS (residual sum of squares). This is simply the sum of the squares of the residuals.
Calculate and print out the TSS (total sum of squares) as well. This is the sum of the squared differences of the outcome variable from its mean.
Then calculate and print the r squared as 1-(RSS/TSS). You should get the same value as was output by your regression.

In [13]:
rss = res.ssr
print('rss: '+ str(rss))

rss: 894.864485408394


In [14]:
tss = np.sum((df['Income']-df['Income'].mean()) ** 2)
print('tss: '+ str(tss))

tss: 12961.992151095286


In [15]:
rsq = 1-(rss/tss)
print('r-squared by calculation: '+ str(rsq))

r-squared by calculation: 0.9309624265331176


In [16]:
print('r-squared by model: '+ str(res.rsquared))

r-squared by model: 0.9309624265331176


### We can see that the Rsquare calculated manually is the same as it is in the model

5. Calculate and print the TSS divided by n-1, where n is the number of observations (which is the number of rows in the original dataframe). This should be the same as one of the values you got earlier. Which one is it the same as?

In [17]:
print('tss divided by n-1: '+ str(tss/(len(df)-1)))

tss divided by n-1: 446.9652465894926


In [18]:
df[['Education','Income']].cov()#.iloc[1,1]

Education      Income
Education  13.271196   74.311836
Income     74.311836  446.965247

### We can see that it is the same as the variance of Income and itself

6. Use the covariance of education and income and the variance of education to calculate and print beta. You should get the same value as you saw in your regression output. 

In [19]:
df[['Education','Income']].cov()#.iloc[0][1]

Education      Income
Education  13.271196   74.311836
Income     74.311836  446.965247

In [20]:
df.var()

Education            13.271196
Income              446.965247
predicted income    416.107851
residual             30.857396
dtype: float64

In [21]:
#beta = 74.311836/13.271196
beta = df[['Education','Income']].cov().iloc[0][1] / df.var()[0]
print('beta: ' + str(beta))

beta: 5.599482872928263


### We can see that this is the same as the coefficient in the ols model we saw before

7. Calculate and print alpha (the intercept of the regression line) as the mean of income minus the following quantity: beta times the mean of education. Use the beta you calculated for the previous question. You should get the same value as you saw in your regression output.

What does this model imply that your income will be if you have zero education? Does that make sense?

In [22]:
alpha = df['Income'].mean()-(beta*df['Education'].mean())
print('alpha: '+ str(alpha))

alpha: -39.44625666018553


### We can see that it is the same as the regression output.  This result implies that at Eduction = 0, you will get -39 income, which obviously does not make sense having negative income. 

8. Create two new standardized columns, education_std and income_std from your education and income columns.
You can do this by first subtracting the mean from each value in each column, and then dividing each value by the standard deviation. This will create new standardized variables with mean equal to zero and standard deviation equal to one. After you do this, double check that the two columns actually have mean zero and standard deviation one by computing and printing these values.
Calculate and print is the correlation between each standardized column and its unstandardized version.

In [23]:
df['education_std'] = (df['Education']-df['Education'].mean())/df['Education'].std()
df['income_std'] = (df['Income']-df['Income'].mean())/df['Income'].std()
df.head()

Education     Income  predicted income   residual  education_std  \
0  10.000000  26.658839         16.548572  10.110267      -1.647010   
1  10.401338  27.306435         18.795856   8.510579      -1.536842   
2  10.842809  22.132410         21.267869   0.864542      -1.415657   
3  11.244147  21.169840         23.515153  -2.345312      -1.305489   
4  11.645485  15.192634         25.762437 -10.569803      -1.195322   

   income_std  
0   -1.110923  
1   -1.080291  
2   -1.325024  
3   -1.370553  
4   -1.653277

In [24]:
print('education mean: '+ str(df['education_std'].mean()))
print('education std: '+ str(df['education_std'].std()))

education mean: 0.0
education std: 1.0


In [25]:
print('income mean: '+ str(round(df['income_std'].mean(),2)))
print('income std: '+ str(df['income_std'].std()))

income mean: -0.0
income std: 1.0


In [26]:
df[['education_std','Education']].corr()

education_std  Education
education_std            1.0        1.0
Education                1.0        1.0

In [27]:
df[['Income','income_std']].corr()

Income  income_std
Income         1.0         1.0
income_std     1.0         1.0

9. Rerun the regression using the standardized versions of the two variables. The r squared should be the same, but the beta will be different. What is it? The beta will be the same as another value you calculated earlier in this assignment. What is it the same as?
Note that alpha should now be zero.

In [28]:
x1 = sm.add_constant(df['education_std'], prepend=False)
mod = sm.OLS(df['income_std'], x1)
res_std = mod.fit()
print(res_std.summary())

                            OLS Regression Results                            
Dep. Variable:             income_std   R-squared:                       0.931
Model:                            OLS   Adj. R-squared:                  0.928
Method:                 Least Squares   F-statistic:                     377.6
Date:                Fri, 05 May 2023   Prob (F-statistic):           8.63e-18
Time:                        23:28:05   Log-Likelihood:                -1.9631
No. Observations:                  30   AIC:                             7.926
Df Residuals:                      28   BIC:                             10.73
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
education_std     0.9649      0.050     19.431

In [29]:
x1 = sm.add_constant(df['Education'], prepend=False)
mod = sm.OLS(df['Income'], x1)
res = mod.fit()
print(res.summary())

                            OLS Regression Results                            
Dep. Variable:                 Income   R-squared:                       0.931
Model:                            OLS   Adj. R-squared:                  0.928
Method:                 Least Squares   F-statistic:                     377.6
Date:                Fri, 05 May 2023   Prob (F-statistic):           8.63e-18
Time:                        23:28:06   Log-Likelihood:                -93.500
No. Observations:                  30   AIC:                             191.0
Df Residuals:                      28   BIC:                             193.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Education      5.5995      0.288     19.431      0.0

In [30]:
print('std rsquared: '+ str(res_std.rsquared))
print('original rsquared: '+ str(res_std.rsquared))
print()
print('std beta: '+ str(res_std.params[0]))
print('original beta: '+ str(res.params[0]))
print('std alpha(should be 0): '+ str(int(res_std.params[1])))

std rsquared: 0.9309624265331176
original rsquared: 0.9309624265331176

std beta: 0.9648639419799657
original beta: 5.599482872928265
std alpha(should be 0): 0


### We can see here that the R-squared is both 0.931. While the standardized data has a beta value of 0.9649, and without standardization, beta = 0.931. 

10. Recompute and print the residuals, RSS, and TSS using the new standardized variables, and then use them to compute and print the r squared. Note that both the TSS and RSS are much lower, but their ratio is the same, and therefore the r squared is the same.
Square the correlation between the two variables and print it. What is it the same as?

In [31]:
res_std.resid

0     0.478218
1     0.402552
2     0.040893
3    -0.110934
4    -0.499954
5    -0.086819
6    -0.617099
7    -0.341561
8     0.090263
9     0.104902
10   -0.250659
11   -0.021043
12    0.026804
13    0.017852
14   -0.031071
15    0.267378
16   -0.101120
17    0.258286
18   -0.036235
19    0.376433
20    0.069451
21    0.172380
22    0.333595
23    0.097343
24    0.191171
25    0.058970
26   -0.202905
27   -0.078899
28   -0.443501
29   -0.164690
dtype: float64

In [32]:
rss= res_std.ssr
print('rss: '+ str(res_std.ssr))

rss: 2.00208963053959


In [33]:
tss = np.sum((df['income_std']-df['income_std'].mean()) ** 2)
print('tss: '+ str(tss))

tss: 29.000000000000007


In [34]:
print('correlation between the two variables: ' + str(df[['education_std','income_std']].corr().iloc[1][0]**2))

correlation between the two variables: 0.9309624265331173


In [35]:
print('r-squared by rss and tss: ' + str(1-rss/tss))

r-squared by rss and tss: 0.9309624265331176


### We can see that the r-squared calculated by rss and tss is the same as the correlation between the two variables